load-test-framework/dotnet_src/Google.PubSub.Flic/Program.cs

```diff
@@ -0,0 +1,174 @@
+using Google.Pubsub.Loadtest;
+using Grpc.Core;
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.PubSub.Flic
+{
+    public class Program
+    {
+        private class Options
+        {
+            [Option("port", Required = false)]
+            public int Port { get; private set; } = 5000;
+        }
+
+        public static async Task<int> Main(string[] args)
+        {
+            var parsed = Parser.Default.ParseArguments<Options>(args);
+            switch (parsed)
+            {
+                case Parsed<Options> success:
+                    // Run the performance test.
+                    await Run(success.Value);
+                    // Do not exit the process after test has completed.
+                    // The client still requires access to complete the perf-test.
+                    // There is no method on the gRPC server instance to wait, so simply
+                    // delay here for a long time.
+                    await Task.Delay(TimeSpan.FromHours(6));
+                    return 0;
+                case NotParsed<Options> failure:
+                    return 1;
+                default:
+                    return 2;
+            }
+        }
+
+        private static async Task Run(Options options)
+        {
+            var startTcs = new TaskCompletionSource<StartRequest>();
+            var stats = new Stats();
+            Console.WriteLine($"Starting dotnet server on port:{options.Port}");
+            var server = new Server
+            {
+                Services = { LoadtestWorker.BindService(new LoadtestServer(startTcs, stats.BuildCheckResponse)) },
+                Ports = { new ServerPort("0.0.0.0", options.Port, ServerCredentials.Insecure) },
+            };
+            server.Start();
+            var startRequest = await startTcs.Task;
+            Console.WriteLine("Start perf test: " +
+                $"Test duration:{(int)startRequest.TestDuration.ToTimeSpan().TotalSeconds}s; " +
+                $"include-ids:{startRequest.IncludeIds}; " +
+                $"CPU scaling:{startRequest.CpuScaling}");
+            var now = DateTime.UtcNow;
+            stats.Init(now, startRequest.IncludeIds);
+            var waitDuration = startRequest.StartTime.ToDateTime() - now;
+            if (waitDuration > TimeSpan.Zero)
+            {
+                await Task.Delay(waitDuration);
+            }
+            switch (startRequest.ClientOptionsCase)
+            {
+                case StartRequest.ClientOptionsOneofCase.PublisherOptions:
+                    await Publish(startRequest, stats);
+                    break;
+                case StartRequest.ClientOptionsOneofCase.SubscriberOptions:
+                default: // no option is sent for subscribe
+                    await Subscribe(startRequest, stats);
+                    break;
+            }
+        }
+
+        private static async Task Publish(StartRequest startRequest, Stats stats)
```

Comment:
Could we move the publisher and subscriber code to separate files, and separate initialization from the action? That way when someone asks "how do I initialize a performant c# publisher", we can point them to the constructor of the publisher, instead of the first section of the publish static method. Something like:

```dotnet
// IWorker.cs
interface IWorker {
  Task Run();
}

// Publisher.cs
internal class Publisher : IWorker {
  public void Init(StartRequest startRequest, Stats stats) {...}
  public async Task Run() {...}
}

// Subscriber.cs
internal class Subscriber : IWorker {
  public void Init(StartRequest startRequest, Stats stats) {...}
  public async Task Run() {...}
}

// Program.cs
IWorker worker;
switch (startRequest.ClientOptionsCase) {
    case StartRequest.ClientOptionsOneofCase.PublisherOptions:
      worker = new Publisher(startRequest, stats);
      break;
    case StartRequest.ClientOptionsOneofCase.SubscriberOptions:
    default:
      worker = new Subscriber(startRequest, stats);
      break;
}
await worker.Run();

```